numba/np/arrayobj.py

```diff
@@ -4958,12 +4958,93 @@ def np_flip(a):
         raise errors.TypingError("Cannot np.flip on %s type" % a)
 
     def impl(a):
-        sl = _build_slice_tuple(a.ndim)
+        sl = _build_flip_slice_tuple(a.ndim)
         return a[sl]
 
     return impl
 
 
+@intrinsic
+def _build_full_slice_tuple(tyctx, sz):
+    """Creates a sz-tuple of full slices"""
+    size = int(sz.literal_value)
+    tuple_type = types.UniTuple(dtype=types.slice2_type, count=size)
+    sig = tuple_type(sz)
+
+    def codegen(context, builder, signature, args):
+        def impl(length, empty_tuple):
+            out = empty_tuple
+            for i in range(length):
+                out = tuple_setitem(out, i, slice(None, None))
+            return out
+
+        inner_argtypes = [types.intp, tuple_type]
+        inner_sig = typing.signature(tuple_type, *inner_argtypes)
+        ll_idx_type = context.get_value_type(types.intp)
+        # Allocate an empty tuple
+        empty_tuple = context.get_constant_undef(tuple_type)
+        inner_args = [ll_idx_type(size), empty_tuple]
+
+        res = context.compile_internal(builder, impl, inner_sig, inner_args)
+        return res
+
+    return sig, codegen
+
+
+@overload(np.array_split)
+def np_array_split(ary, indices_or_sections, axis=0):
+    # If this statement is put at the top of the file, numba fails to import
+    from numba.typed import List
+
+    if ary.ndim > 1:
+        a_type = types.Array(ary.dtype, ary.ndim, "A")
+    else:
+        a_type = typeof(ary)
```

Comment:
it'd probably be a good idea to employ `type_can_asarray` to assert that `ary` can be interpreted as an array, and then `ary = np.asarray(ary)` on the first line of the implementation.

Reply:
I'd like to avoid the `np.asarray` at the beginning if possible, since it could lead to making full copies of array-likes when that's not needed. Also, numpy's `np.split` throws an error for some of the types that pass `type_can_asarray`.

Reply:
I'm not sure what you mean by:
>  it could lead to making full copies of array-likes when that's not needed

IIRC `np.asarray` doesn't make a copy if the argument is already an array. 

> Also, numpy's np.split throws an error for some of the types that pass type_can_asarray

I see, thanks for clarifying, in which case explicit guards for the types that are accepted ought to be added to prevent spurious typing problems emanating?

Reply:
> IIRC np.asarray doesn't make a copy if the argument is already an array.

Yes, but this function should work fine on "array likes". For example, `np.array_split` will return xarray `DataArray`s if you call it on a `DataArray` and `DataFrame`s if called on a `DataFrame`. If this called `np.asarray` on something similar it would return `np.ndarrays`.

> explicit guards for the types that are accepted

Actually, looking at this again, it seems like the cases where I was getting errors might be bugs in numpy. For example:

```python
np.split((1, 2, 3, 4), 2)  # This throws an AttributeError
```

But these should be equivalent, and work (returning lists of arrays)

```python
np.split((1, 2, 3, 4), [2])
np.array_split((1, 2, 3, 4), 2)
```

I think the logic for type conversions is whatever `np.swapaxes` returns.

I guess I would be fine with coercing things to arrays then. I would like it if there was a way for "array likes" to not be converted.

Reply:
Can I know from the type information what attributes/ properties an instance will have?